README.md
## vitaAirHockey v0.2
Air Hockey game for PS Vita.

## Features
The simplest Air Hockey game possible:
- 2 paddles;
- 1 puck;
- The score;
- No AI, local multiplayer only (on the same device).

But... with:
- Multi-touch support (front panel);
- Both analogs support;
- Possibility to increase and decrease the speed in game (with R and L).

## To-do list
- Add IA;
- Add textures;
- Add sounds and music;
- Add multilanguage support;
- Add back touch panel support;

## WIP
- Improve the physics and the collisions; (WIP)
- Add menu; (WIP)
- Optimizing the code; (WIP)

## Done
- Add livearea and icon; (Done)
- Improve the UI. (Done)

## Credits
Thanks to xerpi for the base script (based on vitahelloworld).
Thanks to Rinnegatamante for his patience.
Thanks to all others who have contributed, directly and indirectly.


changelog.txt
## vitaAirHockey v0.1
Air Hockey game for PS Vita.

## Features
The simplest Air Hockey game possible:
- 2 paddles;
- 1 puck;
- The score;
- No AI, local multiplayer only (on the same device).

But... with:
- Multi-touch support (front panel);
- Both analogs support;
- Possibility to increase and decrease the speed in game (with R and L).
________________________________________________________________________

## vitaAirHockey v0.2
- Added icon and livearea images;
- Added minimalist(just 1 choice) "blinking" menu;
- Improved UI;
- Improved the physics and the fun!
- A little code optimization.


v0.1/draw.h

#ifndef DRAW_H
#define DRAW_H

#include <psp2/types.h>

//#define RGBA8(r, g, b, a)      ((((a)&0xFF)<<24) | (((b)&0xFF)<<16) | (((g)&0xFF)<<8) | (((r)&0xFF)<<0))

#define SCREEN_W 960
#define SCREEN_H 544

#define RED   RGBA8(255, 0,   0,   255)
#define GREEN RGBA8(0,   255, 0,   255)
#define BLUE  RGBA8(0,   0,   255, 255)
#define CYAN  RGBA8(0,   255, 255, 255)
#define LIME  RGBA8(50,  205, 50,  255)
#define PURP  RGBA8(147, 112, 219, 255)
#define WHITE RGBA8(255, 255, 255, 255)
#define BLACK RGBA8(0,   0,   0,   255)
#define GREY RGBA8(127, 127, 127, 255)
#define DARK_GREY RGBA8(200, 200, 200, 255)

void init_video();
void end_video();
void swap_buffers();
void clear_screen();
void draw_pixel(uint32_t x, uint32_t y, uint32_t color);
void draw_rectangle(uint32_t x, uint32_t y, uint32_t w, uint32_t h, uint32_t color);
void draw_circle(uint32_t x, uint32_t y, uint32_t radius, uint32_t color);

void font_draw_char(int x, int y, uint32_t color, char c);
void font_draw_string(int x, int y, uint32_t color, const char *string);
void font_draw_stringf(int x, int y, uint32_t color, const char *s, ...);


#endif


source/draw.c

#include <stdio.h>
#include <string.h>
#include <stdarg.h>
#include <psp2/display.h>
#include <psp2/gxm.h>
#include <psp2/kernel/sysmem.h>
#include "draw.h"
#include "utils.h"

extern const unsigned char msx_font[];

static SceDisplayFrameBuf fb[2];
static SceUID fb_memuid[2];
static int cur_fb = 0;

void *gpu_alloc(SceKernelMemBlockType type, unsigned int size, unsigned int attribs, SceUID *uid)
{
	void *mem;

	if (type == SCE_KERNEL_MEMBLOCK_TYPE_USER_CDRAM_RW) {
		size = align_mem(size, 256*1024);
	} else {
		size = align_mem(size, 4*1024);
	}

	*uid = sceKernelAllocMemBlock("gpu_mem", type, size, NULL);

	if (sceKernelGetMemBlockBase(*uid, &mem) < 0)
		return NULL;

	if (sceGxmMapMemory(mem, size, attribs) < 0)
		return NULL;

	return mem;
}

void gpu_free(SceUID uid)
{
	void *mem = NULL;
	if (sceKernelGetMemBlockBase(uid, &mem) < 0)
		return;
	sceGxmUnmapMemory(mem);
	sceKernelFreeMemBlock(uid);
}

void init_video()
{
	int ret;

	SceGxmInitializeParams params;

	params.flags                        = 0x0;
	params.displayQueueMaxPendingCount  = 0x2; //Double buffering
	params.displayQueueCallback         = 0x0;
	params.displayQueueCallbackDataSize = 0x0;
	params.parameterBufferSize          = (16 * 1024 * 1024);

	/* Initialize the GXM */
	ret = sceGxmInitialize(&params);
	printf("sceGxmInitialize(): 0x%08X\n", ret);

	/* Setup framebuffers */
	fb[0].size        = sizeof(fb[0]);
	fb[0].pitch       = SCREEN_W;
	fb[0].pixelformat = SCE_DISPLAY_PIXELFORMAT_A8B8G8R8;
	fb[0].width       = SCREEN_W;
	fb[0].height      = SCREEN_H;

	fb[1].size        = sizeof(fb[1]);
	fb[1].pitch       = SCREEN_W;
	fb[1].pixelformat = SCE_DISPLAY_PIXELFORMAT_A8B8G8R8;
	fb[1].width       = SCREEN_W;
	fb[1].height      = SCREEN_H;

	/* Allocate memory for the framebuffers */
	fb[0].base = gpu_alloc(SCE_KERNEL_MEMBLOCK_TYPE_USER_CDRAM_RW,
		SCREEN_W * SCREEN_H * 4, SCE_GXM_MEMORY_ATTRIB_RW, &fb_memuid[0]);

	if (fb[0].base == NULL) {
		printf("Could not allocate memory for fb[0]. %p", fb[0].base);
		return;
	}

	fb[1].base = gpu_alloc(SCE_KERNEL_MEMBLOCK_TYPE_USER_CDRAM_RW,
		SCREEN_W * SCREEN_H * 4, SCE_GXM_MEMORY_ATTRIB_RW, &fb_memuid[1]);

	if (fb[1].base == NULL) {
		printf("Could not allocate memory for fb[1]. %p", fb[1].base);
		return;
	}

	/* Display the framebuffer 0 */
	cur_fb = 0;
	swap_buffers();

	printf(
		"\nframebuffer 0:\n"
		"\tsize:           0x%08X\n"
		"\tbase:           0x%08X\n"
		"\tpitch:          0x%08X\n"
		"\tpixelformat:    0x%08X\n"
		"\twidth:          0x%08X\n"
		"\theight          0x%08X\n",
		fb[0].size, (uintptr_t)fb[0].base,
		fb[0].pitch, fb[0].pixelformat, fb[0].width, fb[0].height);

	printf(
		"\nframebuffer 1:\n"
		"\tsize:           0x%08X\n"
		"\tbase:           0x%08X\n"
		"\tpitch:          0x%08X\n"
		"\tpixelformat:    0x%08X\n"
		"\twidth:          0x%08X\n"
		"\theight          0x%08X\n",
		fb[1].size, (uintptr_t)fb[1].base,
		fb[1].pitch, fb[1].pixelformat, fb[1].width, fb[1].height);
}

void end_video()
{
	gpu_free(fb_memuid[0]);
	gpu_free(fb_memuid[1]);
	sceGxmTerminate();
}

void swap_buffers()
{
	sceDisplaySetFrameBuf(&fb[cur_fb], SCE_DISPLAY_SETBUF_NEXTFRAME);
	cur_fb ^= 1;
}

void clear_screen()
{
	memset(fb[cur_fb].base, 0xFF, SCREEN_W*SCREEN_H*4);
}

void draw_pixel(uint32_t x, uint32_t y, uint32_t color)
{
	((uint32_t *)fb[cur_fb].base)[x + y*fb[cur_fb].pitch] = color;
}

void draw_rectangle(uint32_t x, uint32_t y, uint32_t w, uint32_t h, uint32_t color)
{
	int i, j;
	for (i = 0; i < h; i++) {
		for (j = 0; j < w; j++) {
			((uint32_t *)fb[cur_fb].base)[(x + j) + (y + i)*fb[cur_fb].pitch] = color;
		}
	}
}

void draw_circle(uint32_t x, uint32_t y, uint32_t radius, uint32_t color)
{
	int r2 = radius * radius;
	int area = r2 << 2;
	int rr = radius << 1;

	int i;
	for (i = 0; i < area; i++) {
		int tx = (i % rr) - radius;
		int ty = (i / rr) - radius;

		if (tx * tx + ty * ty <= r2) {
			draw_pixel(x + tx, y + ty, color);
		}
	}
}

void font_draw_char(int x, int y, uint32_t color, char c)
{
	unsigned char *font = (unsigned char *)(msx_font + (c - (uint32_t)' ') * 8);
	int i, j, pos_x, pos_y;
	for (i = 0; i < 8; ++i) {
		pos_y = y + i*2;
		for (j = 0; j < 8; ++j) {
			pos_x = x + j*2;
			if ((*font & (128 >> j))) {
				draw_pixel(pos_x + 0, pos_y + 0, color);
				draw_pixel(pos_x + 1, pos_y + 0, color);
				draw_pixel(pos_x + 0, pos_y + 1, color);
				draw_pixel(pos_x + 1, pos_y + 1, color);
			}
		}
		++font;
	}
}

void font_draw_string(int x, int y, uint32_t color, const char *string)
{
	if (string == NULL) return;

	int startx = x;
	const char *s = string;

	while (*s) {
		if (*s == '\n') {
			x = startx;
			y += 16;
		} else if (*s == ' ') {
			x += 16;
		} else if(*s == '\t') {
			x += 16*4;
		} else {
			font_draw_char(x, y, color, *s);
			x += 16;
		}
		++s;
	}
}

void font_draw_stringf(int x, int y, uint32_t color, const char *s, ...)
{
	char buf[256];
	va_list argptr;
	va_start(argptr, s);
	vsnprintf(buf, sizeof(buf), s, argptr);
	va_end(argptr);
	font_draw_string(x, y, color, buf);
}



source/menu.c
#include <stdio.h>
#include <stdlib.h>

#include <psp2/ctrl.h>
#include <psp2/touch.h>
#include <psp2/display.h>
#include <psp2/gxm.h>
#include <psp2/types.h>
#include <psp2/moduleinfo.h>
#include <psp2/kernel/processmgr.h>
#include <psp2/ctrl.h>
#include <vita2d.h>

#include "draw.h"
#include "utils.h"

const char *MAIN_MENU[3] = {"Single player(WIP)", "Multiplayer", "Settings(WIP)"};

int i = 0;

SceCtrlData pad;

void main_menu () {
	
	vita2d_set_clear_color(BLACK);
/*
	sceCtrlSetSamplingMode(SCE_CTRL_MODE_ANALOG);
	SceCtrlData pad;
*/
	int j = 0;

	for (i=0; i<=2; i++){
		if (i==1)font_draw_stringf(SCREEN_W/3, SCREEN_H/3 + j, WHITE, "%s\n", MAIN_MENU[i]);
		else font_draw_stringf(SCREEN_W/3, SCREEN_H/3 + j, GREY, "%s\n", MAIN_MENU[i]);
		j = j + 30;
	}
	
	font_draw_stringf(SCREEN_W/3-20, SCREEN_H/2 + j+30, WHITE, "Press X or O to begin");
	
	
/*		if (state >= 3) state = 0;
		if (state <= -1) state = 2;

		for (i=0; i<=2; i++) {
			if (i == state){
				font_draw_stringf(SCREEN_W/3, SCREEN_H/3 + j, WHITE, "%s\n", MAIN_MENU[i]);
			}
			else {
				font_draw_stringf(SCREEN_W/3, SCREEN_H/3 + j, GREY, "%s\n", MAIN_MENU[i]);
			}
			j = j + 30;
		}
		j = 0;
*/
//		sceCtrlPeekBufferPositive(0, &pad, 1);
/*		
		if (pad.buttons & SCE_CTRL_DOWN) {
			state = state+1;
			delay (150000);
		}
		else if (pad.buttons & SCE_CTRL_UP) {
			state = state-1;
			delay (150000);
		}
*///		if (pad.buttons & SCE_CTRL_CROSS) break;
//		i = state;
		
}




v0.1/main.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <malloc.h>
#include <sys/time.h>

#include <psp2/ctrl.h>
#include <psp2/touch.h>
#include <psp2/display.h>
#include <psp2/gxm.h>
#include <psp2/types.h>
#include <psp2/moduleinfo.h>
#include <psp2/kernel/processmgr.h>

#include <vita2d.h>

#include "utils.h"
#include "draw.h"


PSP2_MODULE_INFO(0, 0, "vitaAirHockey");

int main()
{
	init_video();

	/* Enable analog stick */
	sceCtrlSetSamplingMode(SCE_CTRL_MODE_ANALOG);

	/* Enable front and (maybe) back touchscreen */
	sceTouchSetSamplingState(SCE_TOUCH_PORT_FRONT, 1);
	sceTouchSetSamplingState(SCE_TOUCH_PORT_BACK, 1);

	/* FPS counting */
	SceUInt64 cur_micros = 0, delta_micros = 0, last_micros = 0;
	uint32_t frames = 0;
	float fps = 0.0f;

	/* Variables */
/*	int x = SCREEN_W/2 - w/2;
	int y = SCREEN_H/2 - h/2;
*/	int speed = 10;
	
	int r = 35;
//	uint32_t color = RGBA8(255, 0, 0, 255);
	
	/* Mallet1 variables */
	int x1 = 70;
	int y1 = SCREEN_H/2 - r/2;
	int start_x1 = 70;
	
	/* Mallet2 variables */
	int x2 = SCREEN_W - 70;
	int y2 = SCREEN_H/2 - r/2;
	int start_x2 = SCREEN_W - 70;
	
	/* Puck variables */
	int x3 = 180;
	int y3 = SCREEN_H/2 - r/2;
	int puck_is_moving_right = 0;
	int puck_is_moving_left = 0;
	int puck_is_moving_up = 0;
	int puck_is_moving_down = 0;
	int start_x = SCREEN_W/2;
	int start_x3_1 = x3;
	int start_x3_2 = SCREEN_W - 180;
	int start_y = y3;
	
	/* Input variables */
	SceCtrlData pad;
	SceTouchData touch;
	signed char lx, ly, rx, ry;
	
	void delay (unsigned int mseconds) {
		clock_t goal = mseconds + clock();
		while (goal > clock());
	}
	
	/* Game variables */
//	int start_game = 0;
	int m1_points = 0;
	int m2_points = 0;
//	int last_point = 0;

	while (1) {
			
		clear_screen();

		/* Read controls and touchscreen */
		sceCtrlPeekBufferPositive(0, &pad, 1);
		sceTouchPeek(0, &touch, 1);

		if (pad.buttons & SCE_CTRL_START) {
			break;
		}

		/* Draw the center line */
		draw_rectangle (SCREEN_W/2-5, 0, 10, SCREEN_H, GREY);
		
		font_draw_string(10, 10, DARK_GREY, "vitaAirHockey");
		font_draw_stringf(SCREEN_W - 160, 10, GREEN, "FPS: %.2f", fps);
		font_draw_stringf(10, 30, DARK_GREY,
			"(%3d, %3d) (%3d, %3d) (%3d, %3d) size: (%d) speed: %d\n", x1, y1, x2, y2, x3, y3, r, speed);
		font_draw_stringf(10, 50, RED, "%3d -", m1_points);
		font_draw_stringf(60, 50, BLUE, " %3d", m2_points);

		/* Move the mallet 1 */
		if (pad.buttons & SCE_CTRL_UP) {
			y1 -= speed;
		} else 	if (pad.buttons & SCE_CTRL_DOWN) {
			y1 += speed;
		}
		if (pad.buttons & SCE_CTRL_LEFT) {
			x1 -= speed;
		} else 	if (pad.buttons & SCE_CTRL_RIGHT) {
			x1 += speed;
		}
		
		/* Increase/decrease the speed of mallets and puck movement */
		if (pad.buttons & SCE_CTRL_LTRIGGER) {
			delay(5000);
			speed--;
			if (speed < 0) speed = 0;
		} else 	if (pad.buttons & SCE_CTRL_RTRIGGER) {
			delay(5000);
			speed++;
			if (speed > 50) speed = 50;
		}
		
/*		if (pad.buttons & SCE_CTRL_CROSS) {
			color = RGBA8(rand()%255, rand()%255, rand()%255, 255);
		}
*/
		/* Update joystick values */
		lx = (signed char)pad.lx - 128;
		ly = (signed char)pad.ly - 128;
		rx = (signed char)pad.rx - 128;
		ry = (signed char)pad.ry - 128;

		/* Move mallet 1 using the left yoystick */
		if (abs(lx) > 50) {
			x1 += speed * lx/50.0f;
		}
		if (abs(ly) > 50) {
			y1 += speed * ly/50.0f;
		}

		/* Move mallet 2 using the right yoystick */
		if (abs(rx) > 50) {
			x2 += speed * rx/50.0f;
		}
		if (abs(ry) > 50) {
			y2 += speed * ry/50.0f;
		}

		/* Resize using the right yoystick */
/*		if (abs(rx) > 50) {
			w += rx/15.0f;
			if (w < 5) {
				w = 5;
			} else if (w > SCREEN_W) {
				w = SCREEN_W;
			}
		}
		if (abs(ry) > 50) {
			h += ry/15.0f;
			if (h < 5) {
				h = 5;
			} if (h > SCREEN_H) {
				h = SCREEN_H;
			}
		}
*/
		/* Move using the touchscreen! */
		if (touch.reportNum > 0) {
			/* Front touchscreen: 1920x1088 */
			if ((lerp(touch.report[0].x, 1920, SCREEN_W) < SCREEN_W/2) && 
			(lerp(touch.report[0].x, 1920, SCREEN_W) >= x1-100) && (lerp(touch.report[0].x, 1920, SCREEN_W) <= x1+100) &&
			(lerp(touch.report[0].y, 1920, SCREEN_W) >= y1-100) && (lerp(touch.report[0].y, 1920, SCREEN_W) <= y1+100)) {
				x1 = lerp(touch.report[0].x, 1920, SCREEN_W);
				y1 = lerp(touch.report[0].y, 1088, SCREEN_H);
				if ((lerp(touch.report[1].x, 1920, SCREEN_W) > SCREEN_W/2) &&
				(lerp(touch.report[1].x, 1920, SCREEN_W) >= x2-100) && (lerp(touch.report[1].x, 1920, SCREEN_W) <= x2+100) &&
				(lerp(touch.report[1].y, 1920, SCREEN_W) >= y2-100) && (lerp(touch.report[1].y, 1920, SCREEN_W) <= y2+100)) {
					x2 = lerp(touch.report[1].x, 1920, SCREEN_W);
					y2 = lerp(touch.report[1].y, 1088, SCREEN_H);
				}
			}
			else if ((lerp(touch.report[0].x, 1920, SCREEN_W) > SCREEN_W/2) &&
			(lerp(touch.report[0].x, 1920, SCREEN_W) >= x2-100) && (lerp(touch.report[0].x, 1920, SCREEN_W) <= x2+100) &&
			(lerp(touch.report[0].y, 1920, SCREEN_W) >= y2-100) && (lerp(touch.report[0].y, 1920, SCREEN_W) <= y2+100)) {
				x2 = lerp(touch.report[0].x, 1920, SCREEN_W);
				y2 = lerp(touch.report[0].y, 1088, SCREEN_H);
				if ((lerp(touch.report[1].x, 1920, SCREEN_W) < SCREEN_W/2) &&
				(lerp(touch.report[1].x, 1920, SCREEN_W) >= x1-100) && (lerp(touch.report[1].x, 1920, SCREEN_W) <= x1+100) &&
				(lerp(touch.report[1].y, 1920, SCREEN_W) >= y1-100) && (lerp(touch.report[1].y, 1920, SCREEN_W) <= y1+100)) {
					x1 = lerp(touch.report[1].x, 1920, SCREEN_W);
					y1 = lerp(touch.report[1].y, 1088, SCREEN_H);
				}
			}
		}
		
		/* Check left and right collisions */
		if (x1 < r) {
			x1 = r;
		} else if ((x1 + r) > (SCREEN_W/2)) {
			x1 = SCREEN_W/2 - r;
		}

		if ((x2 - r) < (SCREEN_W/2)) {
			x2 = SCREEN_W/2 + r;
		} else if ((x2 + r) > SCREEN_W) {
			x2 = SCREEN_W - r;
		}
		

		/* Check top and bottom collisions */
		if (y1 < r) {
			y1 = r;
		} else if ((y1 + r) > SCREEN_H) {
			y1 = SCREEN_H - r;
		}

		if (y2 < r) {
			y2 = r;
		} else if ((y2 + r) > SCREEN_H) {
			y2 = SCREEN_H - r;
		}
		

		/* Check puck collision and define the movements */
		if (x1 > x3) { // if mallet1 is on the right
			if (((x1-x3) >= 0) && ((x1-x3) <= 70)) {
				if (y1 > y3) { // if mallet1 is on the bottom
					if (((y1-y3) >= 0) && ((y1-y3) <= 70)) {
						puck_is_moving_right=0;
						puck_is_moving_left=1;
						puck_is_moving_up=1;
						puck_is_moving_down=0;
					}
				}
				else if (y1 < y3) { // if mallet1 is on the top
					if (((y3-y1) >= 0) && ((y3-y1) <= 70)) {
						puck_is_moving_right=0;
						puck_is_moving_left=1;
						puck_is_moving_up=0;
						puck_is_moving_down=1;
					}					
				}
				else {
					puck_is_moving_right=0;
					puck_is_moving_left=1;
					puck_is_moving_up=0;
					puck_is_moving_down=0;
				}
			}
		}
		else if (x1 < x3) { //if mallet1 is on the left
			if (((x3-x1) >= 0) && ((x3-x1) <= 70)) { 
				if (y1 > y3) { // if mallet1 is on the bottom
					if (((y1-y3) >= 0) && ((y1-y3) <= 70)) {
						puck_is_moving_right=1;
						puck_is_moving_left=0;
						puck_is_moving_up=1;
						puck_is_moving_down=0;
					}
				}
				else if (y1 < y3) {
					if (((y3-y1) >= 0) && ((y3-y1) <= 70)) { // if mallet1 is on the top
						puck_is_moving_right=1;
						puck_is_moving_left=0;
						puck_is_moving_up=0;
						puck_is_moving_down=1;
					}					
				}
				else {
					if ((puck_is_moving_right == 1) || (puck_is_moving_left == 1)) {
						x3 = x1 + r * 2;
					}
					if ((puck_is_moving_left == 0) && (puck_is_moving_right == 0) && (puck_is_moving_up == 0) && (puck_is_moving_down == 0)) {
						x3 = x1 + r * 2;
					}
						puck_is_moving_right=1;
						puck_is_moving_left=0;
						puck_is_moving_up=0;
						puck_is_moving_down=0;
				}
			}
		}
		else {
			if (y1 > y3) { // if mallet1 is on the bottom
				if (((y1-y3) >= 0) && ((y1-y3) <= 70)) {
					puck_is_moving_right=0;
					puck_is_moving_left=0;
					puck_is_moving_up=1;
					puck_is_moving_down=0;
				}
			}
			else if (y1 < y3) {
				if (((y3-y1) >= 0) && ((y3-y1) <= 70)) { // if mallet1 is on the top
					puck_is_moving_right=0;
					puck_is_moving_left=0;
					puck_is_moving_up=0;
					puck_is_moving_down=1;
				}					
			}
			else {
					puck_is_moving_right=0;
					puck_is_moving_left=0;
					puck_is_moving_up=0;
					puck_is_moving_down=0;
			}
		}

		if (x2 > x3) { // if mallet2 is on the right
			if (((x2-x3) >= 0) && ((x2-x3) <= 70)) {
				if (y2 > y3) { // if mallet2 is on the bottom
					if (((y2-y3) >= 0) && ((y2-y3) <= 70)) {
						puck_is_moving_right=0;
						puck_is_moving_left=1;
						puck_is_moving_up=1;
						puck_is_moving_down=0;
					}
				}
				else if (y2 < y3) { // if mallet2 is on the top
					if (((y3-y2) >= 0) && ((y3-y2) <= 70)) {
						puck_is_moving_right=0;
						puck_is_moving_left=1;
						puck_is_moving_up=0;
						puck_is_moving_down=1;
					}					
				}
				else {
					if ((puck_is_moving_left == 1) || (puck_is_moving_right == 1)) {
						x3 = x2 - r * 2;
					}
					if ((puck_is_moving_left == 0) && (puck_is_moving_right == 0) && (puck_is_moving_up == 0) && (puck_is_moving_down == 0)) {
						x3 = x2 - r * 2;
					}
					puck_is_moving_right=0;
					puck_is_moving_left=1;
					puck_is_moving_up=0;
					puck_is_moving_down=0;
				}
			}
		}
		else if (x2 < x3) { //if mallet2 is on the left
			if (((x3-x2) >= 0) && ((x3-x2) <= 70)) { 
				if (y2 > y3) { // if mallet2 is on the bottom
					if (((y2-y3) >= 0) && ((y2-y3) <= 70)) {
						puck_is_moving_right=1;
						puck_is_moving_left=0;
						puck_is_moving_up=1;
						puck_is_moving_down=0;
					}
				}
				else if (y2 < y3) {
					if (((y3-y2) >= 0) && ((y3-y2) <= 70)) { // if mallet2 is on the top
						puck_is_moving_right=1;
						puck_is_moving_left=0;
						puck_is_moving_up=0;
						puck_is_moving_down=1;
					}					
				}
				else {
						puck_is_moving_right=1;
						puck_is_moving_left=0;
						puck_is_moving_up=0;
						puck_is_moving_down=0;
				}
			}
		}
		else {
			if (y2 > y3) { // if mallet2 is on the bottom
				if (((y2-y3) >= 0) && ((y2-y3) <= 70)) {
					puck_is_moving_right=0;
					puck_is_moving_left=0;
					puck_is_moving_up=1;
					puck_is_moving_down=0;
				}
			}
			else if (y2 < y3) {
				if (((y3-y2) >= 0) && ((y3-y2) <= 70)) { // if mallet2 is on the top
					puck_is_moving_right=0;
					puck_is_moving_left=0;
					puck_is_moving_up=0;
					puck_is_moving_down=1;
				}					
			}
			else {
					puck_is_moving_right=0;
					puck_is_moving_left=0;
					puck_is_moving_up=0;
					puck_is_moving_down=0;
			}
		}
		
		
		if (puck_is_moving_right == 1) {
			if ((x3 + r) < SCREEN_W) {
				x3 += speed;
			}
			else {
				puck_is_moving_right = 0;
				puck_is_moving_left = 1;
			}
		}
		
		if (puck_is_moving_left == 1) {
			if ((x3 - r) > 0) {
				x3 -= speed;
			}
			else {
				puck_is_moving_left = 0;
				puck_is_moving_right = 1;
			}
		}
		
		if (puck_is_moving_up == 1) {
			if ((y3 - r) > (speed)) {
				y3 -= speed;
			}
			else {
				puck_is_moving_up = 0;
				puck_is_moving_down = 1;
			}
		}
		
		if (puck_is_moving_down == 1) {
			if ((y3 + r) <= (SCREEN_H - speed)) {
				y3 += speed;
			}
			else {
				puck_is_moving_down = 0;
				puck_is_moving_up = 1;
			}
		}
		
		if (((x3 + r) <0) || ((x3 - r) > SCREEN_W) || ((y3 + r) < 0) || ((y3 - r) > SCREEN_H)) {
			x3 = start_x;
			y3 = start_y;
			puck_is_moving_up = 0;
			puck_is_moving_down = 0;
			puck_is_moving_left= 0;
			puck_is_moving_right = 0;
		}
		
		if (((x3 + r) >= SCREEN_W-5) && (y3 >= (SCREEN_H/4)) && (y3 <= ((SCREEN_H/4)*3))) {
			m1_points++;
			x3 = start_x3_2;
			y3 = start_y;
			x1 = start_x1;
			y1 = start_y;
			x2 = start_x2;
			y2 = start_y;
			puck_is_moving_up = 0;
			puck_is_moving_down = 0;
			puck_is_moving_left= 0;
			puck_is_moving_right = 0;
		}
		
		if (((x3 - r) <= 5) && (y3 >= (SCREEN_H/4)) && (y3 <= ((SCREEN_H/4)*3))) {
			m2_points++;
			x3 = start_x3_1;
			y3 = start_y;
			x1 = start_x1;
			y1 = start_y;
			x2 = start_x2;
			y2 = start_y;
			puck_is_moving_up = 0;
			puck_is_moving_down = 0;
			puck_is_moving_left= 0;
			puck_is_moving_right = 0;
		}
		
		/* Draw goals */
		draw_rectangle (0, SCREEN_H/4, 5, SCREEN_H/2, GREY); //Left goal
		draw_rectangle (SCREEN_W-5, SCREEN_H/4, 5, SCREEN_H/2, GREY); //Right goal
		
		/* Draw the puck */
		draw_circle(x3, y3, r, BLACK);
		
		/* Draw the mallets */
		draw_circle(x1, y1, r, RED); //Mallet 1 (left)
		draw_circle(x2, y2, r, BLUE); //Mallet 2 (right)

		
		/* Draw a circle */
//		draw_circle(SCREEN_W / 2, SCREEN_H / 2, 50, RGBA8(0,0,255,255));

		/* Calculate FPS */
		cur_micros = sceKernelGetProcessTimeWide();

		if (cur_micros >= (last_micros + 1000000)) {
			delta_micros = cur_micros - last_micros;
			last_micros = cur_micros;
			fps = (frames/(double)delta_micros)*1000000.0f;
			frames = 0;
		}

		swap_buffers();
		sceDisplayWaitVblankStart();
		frames++;
	}

	end_video();
	sceKernelExitProcess(0);
	return 0;
}
